File: src/utils_cpp/struct/Atom.cpp
#include "Atom.hpp"

Atom::Atom() {}
Atom::Atom(const Atom& lhs) : mol_id(lhs.mol_id), mol_name(lhs.mol_name), id(lhs.id), name(lhs.name), xyz(lhs.xyz) {}

Atom& Atom::operator=(const Atom& lhs) {
    Atom t(lhs);
    std::swap(mol_id, t.mol_id);
    std::swap(mol_name, t.mol_name);
    std::swap(id, t.id);
    std::swap(name, t.name);
    std::swap(xyz, t.xyz);

    return *this;
}

Atom::Atom(Atom&& rhs) : mol_id(rhs.mol_id), mol_name(rhs.mol_name), id(rhs.id), name(rhs.name), xyz(rhs.xyz) {
    rhs.mol_id = 1;
    rhs.mol_name = "MOL";
    rhs.id = 1;
    rhs.name = "ATOM";
    rhs.xyz = vec();
}

Atom& Atom::operator=(Atom&& rhs) {
    Atom t(std::move(rhs));
    std::swap(mol_id, t.mol_id);
    std::swap(mol_name, t.mol_name);
    std::swap(id, t.id);
    std::swap(name, t.name);
    std::swap(xyz, t.xyz);

    return *this;
}

File: src/utils_py/gro/Atom.py
from dataclasses import dataclass, field
import numpy as np

@dataclass
class Atom:
    mol_id: int = 1
    mol_name: str = 'MOL'
    name: str = 'ATOM'
    id: int = 1
    xyz: np.array = field(default_factory=np.array)

    def make_gro_line(self) -> str:
        start = f'{self.mol_id:5d}{self.mol_name:<5}{self.name:>5}{self.id%100000:5d}'
        coords = ''.join([f'{x:8.3f}' for x in self.xyz])
        return f'{start}{coords}'

    def copy(self):
        return Atom(
            mol_id = self.mol_id,
            mol_name = self.mol_name,
            name = self.name,
            id = self.id,
            xyz = self.xyz.copy()
        )


@dataclass
class Atom_label:
    mol_id: int = 1
    mol_name: str = 'MOL'
    name: str = 'ATOM'
    id: int = 1

    def copy(self):
        return Atom_label(
            mol_id = self.mol_id,
            mol_name = self.mol_name,
            name = self.name,
            id = self.id
        )

